diff --git a/src/typeinfer.jl b/src/typeinfer.jl
index 40318af..67d9cd9 100644
--- a/src/typeinfer.jl
+++ b/src/typeinfer.jl
@@ -1,8 +1,21 @@
 function typeinf(interp::CustomInterpreter, frame::InferenceState)
+    # debug info before typeinf
+    depth = interp.depth[]
+    io = stdout::IO
+    color = RAIL_COLORS[(depth+1)%N_RAILS+1]
+    print_rails(io, depth)
+    printstyled(io, "┌ @ "; color)
+    println(io, frame.linfo)
+
     interp.depth[] += 1
     ret = @invoke typeinf(interp::AbstractInterpreter, frame::InferenceState)
     interp.depth[] -= 1
 
+    # debug info after typeinf
+    print_rails(io, depth)
+    printstyled(io, "└─> "; color)
+    printlnstyled(io, frame.src.rettype; color = TYPE_ANNOTATION_COLOR)
+
     return ret
 end
 
